janusgraph-core/src/main/java/org/janusgraph/diskstorage/configuration/builder/ReadConfigurationBuilder.java

```diff
@@ -0,0 +1,303 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.configuration.builder;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import org.janusgraph.core.JanusGraphConfigurationException;
+import org.janusgraph.core.JanusGraphException;
+import org.janusgraph.diskstorage.Backend;
+import org.janusgraph.diskstorage.configuration.*;
+import org.janusgraph.diskstorage.configuration.backend.KCVSConfiguration;
+import org.janusgraph.diskstorage.configuration.converter.ReadConfigurationConverter;
+import org.janusgraph.diskstorage.keycolumnvalue.KeyColumnValueStoreManager;
+import org.janusgraph.diskstorage.keycolumnvalue.StoreFeatures;
+import org.janusgraph.diskstorage.util.time.TimestampProviders;
+import org.janusgraph.graphdb.configuration.JanusGraphConstants;
+import org.janusgraph.graphdb.configuration.validator.CompatibilityValidator;
+import org.janusgraph.graphdb.database.management.ManagementSystem;
+import org.janusgraph.util.system.LoggerUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.*;
+
+/**
+ * Builder to build {@link ReadConfiguration} instance of global configuration
+ */
+public class ReadConfigurationBuilder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReadConfigurationBuilder.class);
+
+    private static final String BACKLEVEL_STORAGE_VERSION_EXCEPTION = "The storage version on the client or server is lower than the storage version of the graph: graph storage version %s vs. client storage version %s when opening graph %s.";
+    private static final String INCOMPATIBLE_STORAGE_VERSION_EXCEPTION = "Storage version is incompatible with current client: graph storage version %s vs. client storage version %s when opening graph %s.";
+
+    public ReadConfiguration buildGlobalConfiguration(ReadConfiguration localConfig,
+                                                      BasicConfiguration localBasicConfiguration,
+                                                      ModifiableConfiguration overwrite,
+                                                      KeyColumnValueStoreManager storeManager,
+                                                      ModifiableConfigurationBuilder modifiableConfigurationBuilder){
```

Comment:
@chupman This solution has its own pros and cons. If we make `new ModifiableConfigurationBuilder()` in this method then we won't be able to use it as a mock when we write tests for `ReadConfiguration.buildGlobalConfiguration` that is why it will be a little bit harder to test it. But on the other side we get rid of the dependency parameter. That is why I don't know if it is better to make `new ModifiableConfigurationBuilder()` in this method or set it as a parameter dependency. If you still think that it is a good idea to create a new instance of `ModifiableConfigurationBuilder` in this method than I will make this change.

Reply:
I haven't gone through looking at the testing yet, but based off when I initially looked into refactoring this class I have a feeling I will probably agree with you. I'll take a second look when I review the tests.

Reply:
My thoughts are that if we leave it as is we will be able to add unit tests for this class like:
if `globalWrite` is not frozen then `globalWrite.freezeConfiguration()` should be called and another similar tests with `globalWrite`. If we move the logic of creating `modifiableConfigurationBuilder` inside this this method (through `new` operator) than we won't be able to mock `modifiableConfigurationBuilder` and that is why we won't be able to mock `globalWrite` so we won't be able to write unit tests which use `globalWrite`.